### PR TITLE
Add Specificity metric

### LINKED ISSUE #106: [New Metric] Specificity
## 🚀 Feature
In addition to Precision and Recall it would be nice to have a Specificity metric.

For the implementation I think it would be enough to make a copy of Recall (class und function) and adapt numerator and denominator in _precision_compute.

### Alternatives
For binary classification Specificity is the same as Recall with 0 as true label. 
For multiclass classification this is not as easy as this though.

### PR DESCRIPTION
**Pull request by arvindmuralie77**
# Before submitting

- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure to update the docs?
- [x] Did you write any new necessary tests?

## What does this PR do?
Fixes #106 .
Added classification and functional for specificity. Also added tests

## PR review
Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by maximsch2** [comment]
Can we make the testing simpler by flipping the labels and predictions and calculating regular recall?

File: torchmetrics/classification/specificity.py [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+        multiclass: Optional[bool] = None,
+        compute_on_step: bool = True,
+        dist_sync_on_step: bool = False,
+        process_group: Optional[Any] = None,
+        dist_sync_fn: Callable = None,
+        multilabel: Optional[bool] = None,  # todo: deprecated, remove in v0.4
+        is_multiclass: Optional[bool] = None,  # todo: deprecated, remove in v0.4
+    ):
+        _deprecation_warn_arg_multilabel(multilabel)
+        multiclass = _deprecation_warn_arg_is_multiclass(is_multiclass, multiclass)
```
Comment (SkafteNicki) [inline torchmetrics/classification/specificity.py:149]: no need to deprecate stuff in a new metric :]
```suggestion

```

File: torchmetrics/classification/specificity.py [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+        mdmc_average: Optional[str] = None,
+        ignore_index: Optional[int] = None,
+        top_k: Optional[int] = None,
+        multiclass: Optional[bool] = None,
+        compute_on_step: bool = True,
+        dist_sync_on_step: bool = False,
+        process_group: Optional[Any] = None,
+        dist_sync_fn: Callable = None,
+        multilabel: Optional[bool] = None,  # todo: deprecated, remove in v0.4
+        is_multiclass: Optional[bool] = None,  # todo: deprecated, remove in v0.4
```
Comment (SkafteNicki) [inline torchmetrics/classification/specificity.py:146]:
```suggestion
```

File: torchmetrics/functional/classification/specificity.py [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+    target: Tensor,
+    average: str = "micro",
+    mdmc_average: Optional[str] = None,
+    ignore_index: Optional[int] = None,
+    num_classes: Optional[int] = None,
+    threshold: float = 0.5,
+    top_k: Optional[int] = None,
+    multiclass: Optional[bool] = None,
+    multilabel: Optional[bool] = None,  # todo: deprecated, remove in v0.4
+    is_multiclass: Optional[bool] = None,  # todo: deprecated, remove in v0.4
```
Comment (SkafteNicki) [inline torchmetrics/functional/classification/specificity.py:54]:
```suggestion

```

File: torchmetrics/functional/classification/specificity.py [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+        >>> preds  = torch.tensor([2, 0, 2, 1])
+        >>> target = torch.tensor([1, 1, 2, 0])
+        >>> specificity(preds, target, average='macro', num_classes=3)
+        tensor(0.3333)
+        >>> specificity(preds, target, average='micro')
+        tensor(0.2500)
+
+    """
+    _deprecation_warn_arg_multilabel(multilabel)
+    multiclass = _deprecation_warn_arg_is_multiclass(is_multiclass, multiclass)
```
Comment (SkafteNicki) [inline torchmetrics/functional/classification/specificity.py:166]:
```suggestion

```

File: torchmetrics/functional/classification/specificity.py [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from typing import Optional
+
+import torch
+from torch import Tensor
+
+from torchmetrics.classification.stat_scores import _reduce_stat_scores
+from torchmetrics.functional.classification.stat_scores import _stat_scores_update
+from torchmetrics.utilities import _deprecation_warn_arg_is_multiclass, _deprecation_warn_arg_multilabel
```
Comment (SkafteNicki) [inline torchmetrics/functional/classification/specificity.py:21]:
```suggestion

```

File: torchmetrics/classification/specificity.py [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from typing import Any, Callable, Optional
+
+import torch
+from torch import Tensor
+
+from torchmetrics.classification.stat_scores import StatScores
+from torchmetrics.functional.classification.specificity import _specificity_compute
+from torchmetrics.utilities import _deprecation_warn_arg_is_multiclass, _deprecation_warn_arg_multilabel
```
Comment (SkafteNicki) [inline torchmetrics/classification/specificity.py:21]:
```suggestion

```

**Review by SkafteNicki — COMMENTED** [review]
Overall LGTM, some extra comments:
* remember to add entry to the `docs/source/references/modules.rst` and `docs/source/references/functional.rst` 
* remember to add entry to `CHANGELOG.md` file

File: CHANGELOG.md [hunk 7]
```diff
@@ -9,6 +9,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ### Added
 
+- Added Specificity metric ([PL^106](https://github.com/PyTorchLightning/metrics/issues/106))
```
Comment (SkafteNicki) [inline CHANGELOG.md:12]:
```suggestion
- Added Specificity metric ([#210](https://github.com/PyTorchLightning/metrics/pull/210))
```

**Comment by SkafteNicki** [comment]
Hi @arvindmuralie77,
Code is looking great. Can I request one more change? We are in the process of adding test for differentiability, so it would be great if you could:
* Add this property to the metric: https://github.com/PyTorchLightning/metrics/blob/263a4c7a651f4c85cf79003d59328c08430020bd/torchmetrics/classification/accuracy.py#L275-L277
* Add the corresponding test: https://github.com/PyTorchLightning/metrics/blob/263a4c7a651f4c85cf79003d59328c08430020bd/tests/classification/test_accuracy.py#L106-L116

File: torchmetrics/functional/classification/specificity.py [hunk 8]
```diff
@@ -0,0 +1,184 @@
...
+        raise ValueError("The `mdmc_average` has to be one of {allowed_mdmc_average}, got {mdmc_average}.")
+
+    if average in ["macro", "weighted", "none", None] and (not num_classes or num_classes < 1):
+        raise ValueError(f"When you set `average` as {average}, you have to provide the number of classes.")
+
+    if num_classes and ignore_index is not None and (not 0 <= ignore_index < num_classes or num_classes == 1):
+        raise ValueError(f"The `ignore_index` {ignore_index} is not valid for inputs with {num_classes} classes")
+
+    reduce = "macro" if average in ["weighted", "none", None] else average
+    tp, fp, tn, fn = _stat_scores_update(
```
Comment (SkafteNicki) [inline torchmetrics/functional/classification/specificity.py:172]:
```suggestion
    _, fp, tn, _ = _stat_scores_update(
```

File: torchmetrics/functional/classification/specificity.py [hunk 9]
```diff
@@ -0,0 +1,184 @@
...
+        reduce=reduce,
+        mdmc_reduce=mdmc_average,
+        threshold=threshold,
+        num_classes=num_classes,
+        top_k=top_k,
+        multiclass=multiclass,
+        ignore_index=ignore_index,
+    )
+
+    return _specificity_compute(tp, fp, tn, fn, average, mdmc_average)
```
Comment (SkafteNicki) [inline torchmetrics/functional/classification/specificity.py:184]:
```suggestion
    return _specificity_compute(fp, tn, average, mdmc_average)
```

File: torchmetrics/functional/classification/specificity.py [hunk 10]
```diff
@@ -0,0 +1,184 @@
...
+
+from torchmetrics.classification.stat_scores import _reduce_stat_scores
+from torchmetrics.functional.classification.stat_scores import _stat_scores_update
+
+
+def _specificity_compute(
+    tp: Tensor,
+    fp: Tensor,
+    tn: Tensor,
+    fn: Tensor,
```
Comment (SkafteNicki) [inline torchmetrics/functional/classification/specificity.py:27]:
```suggestion
    fp: Tensor,
    tn: Tensor,
```

File: torchmetrics/functional/classification/specificity.py [hunk 11]
```diff
@@ -0,0 +1,184 @@
...
+    average: str,
+    mdmc_average: Optional[str],
+) -> Tensor:
+    # todo: `tp` is unused
+    # todo: `fn` is unused
```
Comment (SkafteNicki) [inline torchmetrics/functional/classification/specificity.py:32]:
```suggestion
```

File: torchmetrics/classification/specificity.py [hunk 12]
```diff
@@ -0,0 +1,176 @@
...
+
+        Return:
+            The shape of the returned tensor depends on the ``average`` parameter
+
+            - If ``average in ['micro', 'macro', 'weighted', 'samples']``, a one-element tensor will be returned
+            - If ``average in ['none', None]``, the shape will be ``(C,)``, where ``C`` stands  for the number
+              of classes
+        """
+        tp, fp, tn, fn = self._get_final_stats()
+        return _specificity_compute(tp, fp, tn, fn, self.average, self.mdmc_reduce)
```
Comment (SkafteNicki) [inline torchmetrics/classification/specificity.py:172]:
```suggestion
        _, fp, tn, _ = self._get_final_stats()
        return _specificity_compute(fp, tn, self.average, self.mdmc_reduce)
```

**Review by SkafteNicki — APPROVED** [review]
LGTM, only change is to remove the unused arguments. I know that other metrics are doing this, but lets try not to do it.

**Review by Borda — APPROVED** [review]
lgtm
